### ISSUE TITLE
An addition of SD(ObjectSpacing) to aim_pp with the aim to compensate high variance stream maps

### ISSUE THREAD

**Issue by kornkaobat** [ppy/osu-performance#108]
Straight forward idea to add pp based upon the value of S.D.(of ObjectSpacing) throughout the map. Need some feedbacks (pros and cons) first before adding some math formulas to work on.

Pros

Rebalance https://osu.ppy.sh/beatmapsets/837869#osu/1754266

Rebalance https://osu.ppy.sh/beatmapsets/303778#osu/680689

Possibly add some nerf to 1-2 consistent jumps from this S.D. algorithm

**Comment by abraker95** [comment]
Needs further explanation how this will help provided example patterns and resultant values by plugging in those patterns into proposed formula(s).

**Comment by kornkaobat** [comment]
@abraker95 Can you point me the real pp calculation code in this github repo?

**Comment by stanriders** [comment]
This whole repo in the pp calculation code. 

Pretty sure your idea will do exactly the opposite of what you want to achieve. If i understand correctly, you want to buff spacing inconsistencies and the current meta is already about inconsistent maps.

**Comment by abraker95** [comment]
The code is not _that_  hard to find and the repo is small. Here you it anyway

https://github.com/ppy/osu-performance/blob/master/src/performance/osu/OsuScore.cpp

**Comment by kornkaobat** [comment]
@stanriders Well I don't think 1,100pp+ scores that were achieved recently comes from a map that is inconsistent in **ObjectSpacing** because they are just basically cross-screen or full triangle vertical jump with the same O.S. for each 1-2-3/1-2-3-4-5 note. My point here is, since https://osu.ppy.sh/beatmapsets/837869#osu/1754266 with inconsistent ObjectSpacing is ranked, we should expect to see more of it and I think it is hard enough to warrant a score of 900+ or even 1000+ pp nomod.

**Comment by stanriders** [comment]
If you compare a regular diff spike jump farm map to the rubik's cube map you're referring to you'll see that farm maps are actually far less consistent in object spacing. Your approach would only work on streams where spacing changes are actually hard to follow. It can't fix jumps and overall promotes more diffspikes instead of consistent difficulty (which **is** harder to execute)

**Comment by kornkaobat** [comment]
@stanriders So we should seperate stream and jump sections then somehow. One possible way is comparing the time between each hitobject and the real distance of each hitobject on screen. We then use this ratio: Pixel/msecs to calculate what would be called a stream or a jump.

**Comment by stanriders** [comment]
It's already been there from the beginning

**Comment by kornkaobat** [comment]
@stanriders So we should only calculate S.D. for stream sections, will that be okay?

**Comment by stanriders** [comment]
I'm sure there will be thousands of examples where it'll break pp, but you're free to try, we can't be sure without data

**Comment by kornkaobat** [comment]
@abraker95 @stanriders 
Map picture link : https://imgur.com/a/XUJpT1X
Case 1:Rubik's Cube 1st irregular stream at 02:49:638 (1,2,3,4,5,6) -  
419,325->329,300:93.407708
329,300->251,348:91.586025
251,348->191,372:64.621978
191,372->131,345:65.795137
131,345->111,283:65.14599
Standard Deviation, σ: 13.396257730901

Count, N:	5
Sum, Σx:	380.556838
Mean, μ:	76.1113676
Variance, σ2: 	179.45972119272

Case 2:Rubik's Cube 1st normal stream fix at 02:49:638 (1,2,3,4,5,6) -
419,325->360,354:65.74192
360,354->296,368:65.513357
296,368->231,375:65.375837
231,375->167,362:65.306967
167,362->116,321:65.436993
Standard Deviation, σ: 0.14982020321759

Count, N:	5
Sum, Σx:	327.375074
Mean, μ:	65.4750148
Variance, σ2: 	0.022446093292159
This (Case 2) should be set as baseline and all formula of this issue should be in the form of 0.14982020321759 which we will denote as BLnum
Any help appreciated towards creating a formula.

**Comment by abraker95** [comment]
So how many notes will the algorithm take into account? Since these examples use 5, I'll assume 5. So now let me evaluate some basic common patterns @ 180 BPM:

**Double taps [xx-xx-x]:**
000 ->333 : 333
333 ->666 : 333
666 ->1333 : 666
1333 ->1666 : 333
1666 -> 2000 : 333
1666 ->2666 : 666
Standard Deviation, σ: 156.97770542341354

**Triple taps case 1 [xxx-xx]:**
000 ->333 : 333
333 ->666 : 333
666 ->999 : 333
1333 ->1666 : 666
1666 -> 2000 : 333
2000 ->2333 : 333
Standard Deviation, σ: 124.10177275123833

**Triple taps case 2 [x-xxx-x]:**
000 ->333 : 333
333 ->999 : 666
999 ->1333 : 333
1333 ->1666 : 333
1666 -> 2000 : 333
2000 ->2666 : 666
Standard Deviation, σ: 156.97770542341354

Double taps are moderately hard, triple taps are easy enough. So this helps how?

**Comment by kornkaobat** [comment]
@abraker95 I am confused, please use X,Y coordinates or display it in osu map editor.

**Comment by kornkaobat** [comment]
@abraker95 I don’t have time to sample more than 5 spacings but it will be the whole Last Klai stream that will be taken into account. Certain fix for bpm problem is using xy=k Hyperbola equation where x=S.D. and y=bpm results in k=value for further calculation.

**Comment by abraker95** [comment]
> @abraker95 I am confused, please use X,Y coordinates or display it in osu map editor.

There are no X,Y coordinates because these examples do not depend on spatial configuration, and these are arbitrary examples. When you are testing formulas, going into editor to make a pattern is cumbersome. You can plugin mathematical descriptions of patterns like I did to simplify and speed up formula validation.

> Certain fix for bpm problem is using xy=k Hyperbola equation where x=S.D. and y=bpm results in k=value for further calculation.

I have no idea what this is saying

**Comment by stanriders** [comment]
> it will be the whole Last Klai stream that will be taken into account.
How are you going to determine where that stream begins and ends? Streams often have sliders on downbeats so you cant just assume "a lot of circles means stream".
You should also try and test your idea on other maps, rubik's cube is quite an extreme one and you shouldn't be balancing everything around just it

**Comment by kornkaobat** [comment]
@stanriders Hitobject pixel distance divided by the time between the notes (msecs) below some threshold value is a stream. 
I don’t have time but here are the examples: Furioso Melodica 2017 VIP, Burnt Rice (Master), Electric Pink, Zutomayo - Humanoid [ Carbonated Water ].

**Comment by kornkaobat** [comment]
And here's the pp formula revision 1: New_pp=final_pp+(MSD/0.14982020321759)
When MSD = Standard Deviation of ObjectSpacing calculated from all stream sections of the map
Decimals of 0.14982020321759 can be omitted if it is resource intensive.

**Comment by kornkaobat** [comment]
Using 6 decimal place, Rubik’s Cube will gain 89.42 or 90 pp to 908 pp

**Comment by stanriders** [comment]
With values like this you better just look at delta's rebalance imo, rubik's cube is already 929pp for ss there

**Comment by kornkaobat** [comment]
@stanriders Who’s delta, I can’t see his opened issue. Is it the statistical rebalance one?

**Comment by stanriders** [comment]
There's no issue yet, but there're: [Incomplete document explaining things](https://docs.google.com/document/d/1rymnSJ35avbWLwFe45twHnB5zM1BIQcj_mV6D4RE-H8), [source code](https://github.com/HeBuwei/osu/tree/master/osu.Game.Rulesets.Osu/Difficulty) and [a web calculator](https://newpp.stanr.info/)

**Comment by kornkaobat** [comment]
After some code digging, I've found out that there's no variable for hit object spacing as from here;
```
        s64 scoreId,
	EGamemode mode,
	s64 userId,
	s32 beatmapId,
	s32 score,
	s32 maxCombo,
	s32 num300,
	s32 num100,
	s32 num50,
	s32 numMiss,
	s32 numGeki,
	s32 numKatu,
	EMods mods,
```
We could tie a new variable known as s32 spacingsdev to s32 beatmapId so that it becomes a constant retrievable from the server when the mapper commits any changes to the map. Only the mapper will need to do spacingsdev calculations on their side and send it to the server. Validity of the value will be signed using developer's PGP key.

**Comment by kornkaobat** [comment]
For spacingsdev calculations, we will parse and process from .osz beatmap file.

**Comment by abraker95** [comment]
I don't think you understand how PP calculation works. The data you just listed gets sent to osu's servers to where pp is calculated. No map data gets sent because the map data is already stored on the servers. Which means `spacingsdev` would be calculated there too.

**Comment by peppy** [comment]
it is definitely something we can add - we have the facilities to store new variables for performance calculation use.

**Comment by joseph-ireland** [comment]
There is a full implementation of difficulty and PP in the main repo https://github.com/ppy/osu/tree/master/osu.Game.Rulesets.Osu/Difficulty, and tools to run it  https://github.com/ppy/osu-tools 

I feel like that should be in the readme somewhere, it's probably what most people are looking for when tehy click this repo

**Comment by kornkaobat** [comment]
@abraker95 I think spacingsdev calculations on the server would put it into heavy load due to the amount of math each beatmap processing needs. If the server can handle it that could be probable.

**Comment by kornkaobat** [comment]
[osuppstdevflowchart.pdf](https://github.com/ppy/osu-performance/files/4238931/osuppstdevflowchart.pdf)

**Comment by kornkaobat** [comment]
The next hard part is now "programming". I'll seek help from some programmers.
By the way, @stanriders to answer the slider problem, we'll use end-to-start as distance and timing for sliders.

**Comment by kornkaobat** [comment]
As for any mods added to plays, use NoMod base as calculation. Then, apply score multiplier to spacingsdev (aka Additional_pp ) before adding to original_pp.

**Comment by joseph-ireland** [comment]
There is already infrastructure to do everything that you want, but the relevant code isn't just in this repo. This repo only contains code to load beatmap difficulty attributes from a database and calculate the PP of a replay. That's why there's nothing about object spacing here. The code that calculates these difficulty attributes and knows about object spacing can be found in the links I posted.

**Comment by kornkaobat** [comment]
@joseph-ireland Then I’ll do a pull request to retrieve more attributes from osu.difficulty repository. public double JumpDistance, public readonly double StrainTime

**Comment by kornkaobat** [comment]
@peppy Sorry to ask, but how did you transfer multiple variable across repositories?

**Comment by joseph-ireland** [comment]
You don't need to just yet, you can do everything in osu-tools and the main osu repo, just install the recommended requirements and follow "getting started" from the osu-tools readme, then you can start editing stuff in osu.Game.Rulesets.Difficulty and see the results. 

This repo is just for use on the server to quickly calculate pp without having to do the hard calculations every time. osu-tools also calculates pp, but calculates map difficulty too, and that part is likely what you want to change.

**Comment by kornkaobat** [comment]
What is applyDiminishingExp(osuCurrent.JumpDistance); do? There seems to be no reference function in the source code repo.

**Comment by VINXIS** [comment]
https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs#L57

**Comment by kornkaobat** [comment]
`// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
// See the LICENCE file in the repository root for full licence text.

using System;
using System.Linq;
using System.Math;
using osu.Game.Rulesets.Difficulty.Preprocessing;
using osu.Game.Rulesets.Difficulty.Skills;
using osu.Game.Rulesets.Osu.Difficulty.Preprocessing;
using osu.Game.Rulesets.Osu.Objects;

namespace osu.Game.Rulesets.Osu.Difficulty.Skills
{
    /// <summary>
    /// Represents the skill required to correctly aim at every object in the map with a uniform CircleSize and normalized distances.
    /// </summary>
    public class Aim : Skill
    {
        private const double angle_bonus_begin = Math.PI / 3;
        private const double timing_threshold = 107;
        private const double streamaimconst = 2.42;
        private const double stdevconst = 0.149820;

        protected override double SkillMultiplier => 26.25;
        protected override double StrainDecayBase => 0.15;

        protected override double StrainValueOf(DifficultyHitObject current)
        {
            if (current.BaseObject is Spinner)
                return 0;

            var osuCurrent = (OsuDifficultyHitObject)current;

            double result = 0;

            if (Previous.Count > 0)
            {
                var osuPrevious = (OsuDifficultyHitObject)Previous[0];

                if (osuCurrent.Angle != null && osuCurrent.Angle.Value > angle_bonus_begin)
                {
                    const double scale = 90;

                    var angleBonus = Math.Sqrt(
                        Math.Max(osuPrevious.JumpDistance - scale, 0)
                        * Math.Pow(Math.Sin(osuCurrent.Angle.Value - angle_bonus_begin), 2)
                        * Math.Max(osuCurrent.JumpDistance - scale, 0));
                    result = 1.5 * applyDiminishingExp(Math.Max(0, angleBonus)) / Math.Max(timing_threshold, osuPrevious.StrainTime);
                }
            }

            double jumpDistanceExp = applyDiminishingExp(osuCurrent.JumpDistance);
            double travelDistanceExp = applyDiminishingExp(osuCurrent.TravelDistance);
            
            /// Standard Deviation Calculation Code courtesy of Roman http://csharphelper.com/blog/2015/12/make-an-extension-method-that-calculates-standard-deviation-in-c/
            public static double GetStandardDeviation(List values)
             {
             double avg = values.Average();
             double sum = values.Sum(v => (v - avg) * (v-avg));
             double denominator = values.Count - 1;
             return denominator > 0.0 ? Math.Sqrt(sum / denominator) : -1;
             };
        
            public static double StreamDeterminant(double deltadist,double deltatime){
               sectionvelocity = deltadist/deltatime
               if (sectionvelocity < streamaimconst) 
                { 
                    finalsd = GetStandardDeviation(osuCurrent.JumpDistance);
                    finalsdpp = finalsd/stdevconst
                 };
            };
            StreamDeterminant(JumpDistance,StrainTime);
            

            return Math.Max(
                result + (jumpDistanceExp + travelDistanceExp + Math.Sqrt(travelDistanceExp * jumpDistanceExp)) / Math.Max(osuCurrent.StrainTime, timing_threshold),
                (Math.Sqrt(travelDistanceExp * jumpDistanceExp) + jumpDistanceExp + travelDistanceExp) / osuCurrent.StrainTime
            ) + finalsdpp;
        }

        private double applyDiminishingExp(double val) => Math.Pow(val, 0.99);
    }
}
`
My programming might be wrong but the general idea is correct. Any feedback appreciated.

**Comment by stanriders** [comment]
Just test it and see the results

**Comment by kornkaobat** [comment]
What is the difference between OsuCurrent.Variable , OsuPrevious.Variable, and normal variable?

**Comment by kornkaobat** [comment]
Seems like this is going to involve SD summation formula for each strain (Previous and Current) section.

**Comment by kornkaobat** [comment]
![48BE7843-C16C-4E50-A696-CC1E07D8FF31](https://user-images.githubusercontent.com/26484978/75610447-c5577100-5b43-11ea-8b29-fca63b3bf389.jpeg)

**Comment by kornkaobat** [comment]
...\aim.cs
In one week this project should be ready for a pull request.
```
// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
// See the LICENCE file in the repository root for full licence text.

using System;
using System.Linq;
using System.Math;
using osu.Game.Rulesets.Difficulty;
using osu.Game.Rulesets.Difficulty.Preprocessing;
using osu.Game.Rulesets.Difficulty.Skills;
using osu.Game.Rulesets.Osu.Difficulty.Preprocessing;
using osu.Game.Rulesets.Osu.Objects;

namespace osu.Game.Rulesets.Osu.Difficulty.Skills
{
    /// <summary>
    /// Represents the skill required to correctly aim at every object in the map with a uniform CircleSize and normalized distances.
    /// </summary>
    public class Aim : Skill
    {
        private const double angle_bonus_begin = Math.PI / 3;
        private const double timing_threshold = 107;
        private const double streamaimconst = 2.42;
        private const double stdevconst = 0.149820;
        double[] JumpDistanceArray;
        int jumpdistpointercount = 0

        var osuCurrent = (OsuDifficultyHitObject)current;
        var osuPrevious = (OsuDifficultyHitObject)Previous[0];

        protected override double SkillMultiplier => 26.25;
        protected override double StrainDecayBase => 0.15;

        /// Standard Deviation Calculation Code courtesy of Roman http://csharphelper.com/blog/2015/12/make-an-extension-method-that-calculates-standard-deviation-in-c/
        /// public static double GetStandardDeviation(List values)
        ///     {
        ///     double avg = values.Average();
        ///     double sum = values.Sum(v => (v - avg) * (v-avg));
        ///     double denominator = values.Count - 1;
        ///     return denominator > 0.0 ? Math.Sqrt(sum / denominator) : -1;
        ///     };
        
        public static double StreamDeterminant(double deltadist,double deltatime)
            {
               sectionvelocity = deltadist/deltatime
               if (sectionvelocity < streamaimconst) 
                { 
                    JumpDistanceArray[jumpdistpointercount] = deltadist;
                    int jumpdistpointercount = jumpdistpointercount + 1
                /// Use JumpDistanceArray to calculate SD
                /// double average = JumpDistanceArray.Average();
                /// double sumOfSquaresOfDifferences = JumpDistanceArray.Select(val => (val - average) * (val - average)).Sum();
                /// double finalsd = Math.Sqrt(sumOfSquaresOfDifferences / JumpDistanceArray.Length); 
                /// double finalsdpp = finalsd/stdevconst;
                };
            };
        

        protected override double StrainValueOf(DifficultyHitObject current)
        {
            if (current.BaseObject is Spinner)
                return 0;

            var osuCurrent = (OsuDifficultyHitObject)current;

            double result = 0;

            StreamDeterminant(osuCurrent.JumpDistance,osuCurrent.StrainTime);

            if (Previous.Count > 0)
            {
                var osuPrevious = (OsuDifficultyHitObject)Previous[0];

                if (osuCurrent.Angle != null && osuCurrent.Angle.Value > angle_bonus_begin)
                {
                    const double scale = 90;

                    var angleBonus = Math.Sqrt(
                        Math.Max(osuPrevious.JumpDistance - scale, 0)
                        * Math.Pow(Math.Sin(osuCurrent.Angle.Value - angle_bonus_begin), 2)
                        * Math.Max(osuCurrent.JumpDistance - scale, 0));
                    result = 1.5 * applyDiminishingExp(Math.Max(0, angleBonus)) / Math.Max(timing_threshold, osuPrevious.StrainTime);
                }
            }

            double jumpDistanceExp = applyDiminishingExp(osuCurrent.JumpDistance);
            double travelDistanceExp = applyDiminishingExp(osuCurrent.TravelDistance);
            
            return Math.Max(
                result + (jumpDistanceExp + travelDistanceExp + Math.Sqrt(travelDistanceExp * jumpDistanceExp)) / Math.Max(osuCurrent.StrainTime, timing_threshold),
                (Math.Sqrt(travelDistanceExp * jumpDistanceExp) + jumpDistanceExp + travelDistanceExp) / osuCurrent.StrainTime
            );
        }

        private double applyDiminishingExp(double val) => Math.Pow(val, 0.99);
    }
}

```

**Comment by kornkaobat** [comment]
I am about to finalize my programming works, do you think I should add the new Standard Deviation pp to totalStrain,aimValue (aimpp) or totalValue (totalpp) ? ( This is mainly a concern because of game modifiers or mods - affecting deltadist and deltatime)

**Comment by kornkaobat** [comment]
Funny thing is: osu-tools is written in C# while this repo is written in C++ so I have to write 2x code. 😑

**Comment by kornkaobat** [comment]
[https://github.com/ppy/osu/pull/8065](Current Pull Request)
Difficulty\Skills\Aim.cs(46,39): error CS0844: Cannot use local variable 'jumpdistpointercount' before it is declared. The declaration of the local variable hides the field 'Aim.jumpdistpointercount'. [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\Skills\Aim.cs(47,25): error CS0136: A local or parameter named 'jumpdistpointercount' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\Skills\Aim.cs(46,21): error CS0165: Use of unassigned local variable 'JumpDistanceArray' [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\Skills\Aim.cs(43,20): warning CS0219: The variable 'jumpdistpointercount' is assigned but its value is never used [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\Skills\Aim.cs(39,30): error CS0161: 'Aim.StreamDeterminant(double, double)': not all code paths return a value [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\OsuPerformanceCalculator.cs(97,30): error CS0103: The name 'AimStrain' does not exist in the current context [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\OsuPerformanceCalculator.cs(98,48): error CS0103: The name 'AimStrain' does not exist in the current context [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\OsuPerformanceCalculator.cs(99,68): error CS0103: The name 'AimStrain' does not exist in the current context [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\OsuPerformanceCalculator.cs(101,17): error CS0029: Cannot implicitly convert type 'int' to 'bool' [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\OsuPerformanceCalculator.cs(102,19): error CS1023: Embedded statement cannot be a declaration or labeled statement [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\OsuPerformanceCalculator.cs(102,26): error CS0136: A local or parameter named 'finalsdpp' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\OsuPerformanceCalculator.cs(102,38): error CS0165: Use of unassigned local variable 'finalsdpp' [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\Skills\Aim.cs(24,18): warning CS0169: The field 'Aim.JumpDistanceArray' is never used [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]
Difficulty\Skills\Aim.cs(25,13): warning CS0414: The field 'Aim.jumpdistpointercount' is assigned but its value is never used [C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\osu.Game.Rulesets.Osu.csproj]

The build failed. Fix the build errors and run again.

**Comment by kornkaobat** [comment]
Anyone donate me .osr and .osu related to that .osr file. I've successfully compiled the program locally.

**Comment by kornkaobat** [comment]
```
C:\Users\Korn\osu-tools\PerformanceCalculator>dotnet run -- profile WhiteCat <API_KEY> --output whitecat.txt
Getting user data...
Getting user top scores...
Downloading 2097898.osu...

Unhandled Exception: System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at osu.Game.Rulesets.Osu.Difficulty.Skills.Aim.StrainValueOf(DifficultyHitObject current) in C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\Difficulty\Skills\Aim.cs:line 59
   at osu.Game.Rulesets.Difficulty.Skills.Skill.Process(DifficultyHitObject current) in C:\Users\Korn\osu-tools\osu\osu.Game\Rulesets\Difficulty\Skills\Skill.cs:line 55
   at osu.Game.Rulesets.Difficulty.DifficultyCalculator.calculate(IBeatmap beatmap, Mod[] mods, Double clockRate) in C:\Users\Korn\osu-tools\osu\osu.Game\Rulesets\Difficulty\DifficultyCalculator.cs:line 93
   at osu.Game.Rulesets.Difficulty.DifficultyCalculator.Calculate(Mod[] mods) in C:\Users\Korn\osu-tools\osu\osu.Game\Rulesets\Difficulty\DifficultyCalculator.cs:line 47
   at osu.Game.Rulesets.Difficulty.PerformanceCalculator..ctor(Ruleset ruleset, WorkingBeatmap beatmap, ScoreInfo score) in C:\Users\Korn\osu-tools\osu\osu.Game\Rulesets\Difficulty\PerformanceCalculator.cs:line 31
   at osu.Game.Rulesets.Osu.Difficulty.OsuPerformanceCalculator..ctor(Ruleset ruleset, WorkingBeatmap beatmap, ScoreInfo score) in C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\Difficulty\OsuPerformanceCalculator.cs:line 40
   at osu.Game.Rulesets.Osu.OsuRuleset.CreatePerformanceCalculator(WorkingBeatmap beatmap, ScoreInfo score) in C:\Users\Korn\osu-tools\osu\osu.Game.Rulesets.Osu\OsuRuleset.cs:line 156
   at PerformanceCalculator.Profile.ProfileCommand.Execute() in C:\Users\Korn\osu-tools\PerformanceCalculator\Profile\ProfileCommand.cs:line 83
   at PerformanceCalculator.ProcessorCommand.OnExecute(CommandLineApplication app, IConsole console) in C:\Users\Korn\osu-tools\PerformanceCalculator\ProcessorCommand.cs:line 26
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at McMaster.Extensions.CommandLineUtils.Conventions.ExecuteMethodConvention.Invoke(MethodInfo method, Object instance, Object[] arguments)
   at McMaster.Extensions.CommandLineUtils.Conventions.ExecuteMethodConvention.<OnExecute>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at McMaster.Extensions.CommandLineUtils.Conventions.ExecuteMethodConvention.<>c__DisplayClass0_0.<<Apply>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.<ExecuteAsync>d__160.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.<ExecuteAsync>d__191`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute[TApp](CommandLineContext context)
   at PerformanceCalculator.Program.Main(String[] args) in C:\Users\Korn\osu-tools\PerformanceCalculator\Program.cs:line 27
```

**Comment by kornkaobat** [comment]
...

**Comment by kornkaobat** [comment]
[WhiteCat's new pp](https://github.com/ppy/osu-performance/files/4291437/message.1.txt)

**Comment by kornkaobat** [comment]
Most comments I've made have been deleted due to me losing hope that this project will stride forward. A lot of temporary white paper is currently stored here : [Osu! Dev Discord](https://discordapp.com/channels/188630481301012481/380598781432823815). More to come soon, especially my repo for anyone to use.

**Comment by kornkaobat** [comment]
https://github.com/kornkaobat/osu-tools and subfolder https://github.com/kornkaobat/osu

**Comment by kornkaobat** [comment]
For all of my old deleted comments see https://gitmemory.com/kornkaobat

**Comment by kornkaobat** [comment]
[idke.txt](https://github.com/ppy/osu-performance/files/4358995/idke.txt)
[gayzmcgee.txt](https://github.com/ppy/osu-performance/files/4358996/gayzmcgee.txt)
[cookiezi.txt](https://github.com/ppy/osu-performance/files/4358997/cookiezi.txt)
[whitecat.txt](https://github.com/ppy/osu-performance/files/4358998/whitecat.txt)
[ryuk.txt](https://github.com/ppy/osu-performance/files/4358999/ryuk.txt)
[vaxei.txt](https://github.com/ppy/osu-performance/files/4359000/vaxei.txt)
Current pp values as of https://github.com/kornkaobat/osu/pull/17
